src/client/mod.rs

```diff
@@ -409,134 +484,136 @@ impl IotHubClient {
-            event_handler: Box::new(event_handler),
+            tx_connection_status,
+            tx_twin_desired,
+            tx_direct_method,
+            tx_incoming_message,
         });
 
         client.set_callbacks()?;
 
+        client.set_options()?;
+
         Ok(client)
     }
 
-    /// Let's you either create an outgoing D2C messages or parse an incoming cloud to device (C2D) messages.
+    /// Call this function to send a message (D2C) to iothub.
     /// ```rust, no_run
-    /// # use azure_iot_sdk::client::*;
-    /// # struct MyEventHandler {}
-    /// # impl EventHandler for MyEventHandler {}
-    /// #
-    /// #
-    /// # let event_handler = MyEventHandler{};
-    /// # let mut client = IotHubClient::from_identity_service(event_handler).unwrap();
-    /// #
-    /// let msg = IotMessage::builder()
-    ///     .set_body(
-    ///         serde_json::to_vec(r#"{"my telemetry message": "hi from device"}"#).unwrap(),
-    ///     )
-    ///     .set_id("my msg id")
-    ///     .set_correlation_id("my correleation id")
-    ///     .set_property(
-    ///         "my property key",
-    ///         "my property value",
-    ///     )
-    ///     .set_output_queue("my output queue")
-    ///     .build()
-    ///     .unwrap();
+    /// use azure_iot_sdk::client::*;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let mut client = IotHubClient::from_identity_service(None, None, None, None).await.unwrap();
     ///
-    /// client.send_d2c_message(msg);
+    ///     let msg = IotMessage::builder()
+    ///         .set_body(
+    ///             serde_json::to_vec(r#"{"my telemetry message": "hi from device"}"#).unwrap(),
+    ///         )
+    ///         .set_id("my msg id")
+    ///         .set_correlation_id("my correleation id")
+    ///         .set_property(
+    ///             "my property key",
+    ///             "my property value",
+    ///         )
+    ///         .set_output_queue("my output queue")
+    ///         .build()
+    ///         .unwrap();
+    ///
+    ///     client.send_d2c_message(msg);
+    /// }
     /// ```
-    pub fn send_d2c_message(&mut self, mut message: IotMessage) -> Result<u32> {
+    async fn send_d2c_message(&mut self, mut message: IotMessage) -> Result<()> {
         let handle = message.create_outgoing_handle()?;
         let queue = message.output_queue.clone();
-        let ctx = rand::thread_rng().gen::<u32>();
+        let notify = Arc::new(Notify::new());
+        let result = Arc::new(Mutex::new(false));
 
-        debug!("send_event with internal id: {}", ctx);
+        debug!("send_d2c_message: {queue:#?}");
 
         self.twin.send_event_to_output_async(
             handle,
             queue,
             Some(IotHubClient::c_d2c_confirmation_callback),
-            ctx,
+            Box::into_raw(Box::new((result.clone(), notify.clone()))) as *mut c_void,
         )?;
 
-        Ok(ctx)
+        notify.notified().await;
```

Comment:
The pattern appears to be: set state var, notify, await notify, read state var. Is the mutex then really necessary (i.e. the access order for the result is guarded by it)? It seems to me that a tokio::oneshot could be sufficient here?